### PR TITLE
added ccd_process to core

### PR DESCRIPTION
**Pull request by crawfordsm**
This is the first pull request for adding a ccd_process task that has all the reductions steps in one function.

### CODE REVIEW

**Comment by crawfordsm** [comment]
Before merging, still need to do the following:
- [ ]  add flatfielding
- [ ] tests for the task
- [ ] documentation for the task

**Comment by crawfordsm** [comment]
@mwcraig okay this ended up being bigger than I thought.  I still need to rebase this and ran into some conflicts that ended up removing your update of sigma_func.  I'll put that back in and push this again.

I've also included a fix in ccddata for masks being left out when simple arithmetic was being done.

**Comment by crawfordsm** [comment]
also i need to add a change log

File: CHANGES.rst [hunk 1]
```diff
@@ -22,6 +22,7 @@ New Features
   ``average_combine`` and ``median_combine``. [#258]
 - Add a new keyword to ImageFileCollection.files_filtered to return the full
   path to a file [#275]
+- Added ccd_process for handling multiply steps. [#211]
```
Comment (mwcraig) [inline CHANGES.rst]: multiply → multiple

File: ccdproc/ccddata.py [hunk 2]
```diff
@@ -343,9 +343,10 @@ def _ccddata_arithmetic(self, other, operation, scale_uncertainty=False):
         else:
             result_uncertainty = None
 
+        new_mask = copy.deepcopy(self.mask)
         new_meta = copy.deepcopy(self.meta)
         new_wcs = copy.deepcopy(self.wcs)
-        result = CCDData(result_data, unit=result_unit,
+        result = CCDData(result_data, unit=result_unit, mask=new_mask,
```
Comment (mwcraig) [inline ccdproc/ccddata.py]: good catch!

File: ccdproc/core.py [hunk 3]
```diff
@@ -45,6 +46,161 @@
...
+def ccd_process(ccd, oscan=None, trim=None, error=False, master_bias=None,
+                master_flat=None, bad_pixel_mask=None, gain=None,
+                readnoise=None, oscan_median=True, oscan_model=None,
+                min_value=None):
+    """Perform basic processing on ccd data.
+
+    The following steps can be included:
+    * overscan correction
+    * trimming of the image
+    * create edeviation frame
```
Comment (mwcraig) [inline ccdproc/core.py]: edeviation → deviation

File: ccdproc/core.py [hunk 4]
```diff
@@ -45,6 +46,161 @@
...
+    * gain correction
+    * add a mask to the data
+    * subtraction of master bias
```
Comment (mwcraig) [inline ccdproc/core.py]: Also dark subtraction and flat correction, I think?

File: ccdproc/core.py [hunk 5]
```diff
@@ -45,6 +46,161 @@
...
+    trim: None or str
+        For no trim correction, set to None.   Otherwise proivde a region
+        of ccd from which the image should be trimmed, using the FITS
+        conventions for index order and index start.
+
+    error: boolean
+        If True, create an uncertainty array for ccd
+
+    master_bias: None, `~numpy.ndarray`,  or `~ccdproc.CCDData`
+        A materbias frame to be subtracted from ccd.
```
Comment (mwcraig) [inline ccdproc/core.py]: materbias → master bias

File: ccdproc/core.py [hunk 6]
```diff
@@ -45,6 +46,161 @@
...
+
+    master_flat: None, `~numpy.ndarray`,  or `~ccdproc.CCDData`
+        A materflat frame to be divided into ccd.
```
Comment (mwcraig) [inline ccdproc/core.py]: materflat → master flat

File: ccdproc/core.py (hunk 4 above)
Comment (mwcraig) [inline ccdproc/core.py]: Well, flat correction at least.

Any reason not to add dark subtraction?

File: ccdproc/core.py [hunk 7]
```diff
@@ -45,6 +46,161 @@
...
+
+    # create the error frame
+    if error and gain is not None and readnoise is not None:
+        nccd = create_deviation(nccd, gain=gain, readnoise=readnoise)
+    elif error and (gain is None or readnoise is None):
+        raise ValueError(
+            'gain and readnoise must be specified to create error frame')
+
+    # apply the bad pixel mask
+    if isinstance(bad_pixel_mask, np.ndarray):
```
Comment (mwcraig) [inline ccdproc/core.py]: should `bad_pixel_mask` be or'ed with `ccd.mask`?

File: ccdproc/core.py [hunk 8]
```diff
@@ -670,14 +826,12 @@ def sigma_func(arr, axis=None):
...
 
-
     Returns
     -------
     float
         uncertainty of array estimated from median absolute deviation.
     """
-
-    return 1.482602218505602 * stats.median_absolute_deviation(arr, axis=axis)
+    return 1.482602218505602 * stats.median_absolute_deviation(arr)
```
Comment (mwcraig) [inline ccdproc/core.py]: This should be reversed.

File: ccdproc/tests/test_ccdproc.py [hunk 9]
```diff
@@ -771,3 +771,71 @@ def test_wcs_project_onto_scale_wcs(ccd_data):
...
+    # master bias check
+    with pytest.raises(TypeError):
+        ccd_process(ccd_data, master_bias=3)
+
+    # master flat check
+    with pytest.raises(TypeError):
+        ccd_process(ccd_data, master_flat=3)
+
+
+def test_ccd_process():
```
Comment (mwcraig) [inline ccdproc/tests/test_ccdproc.py]: Could you add a test in which either the master bias or flat is not a CCDData to make sure that works? Not critical...I just don't remember whether arithmetic works with a non-CCDData array.

**Comment by mwcraig** [comment]
Looks good; main substantive suggestion is to allow for dark subtraction.

File: ccdproc/tests/test_ccdproc.py (hunk 9 above)
Comment (crawfordsm) [inline ccdproc/tests/test_ccdproc.py]: It is a problem and I should have made a note of it.   I think it is outside of this pull request but I'll open an issue about it.

Comment (crawfordsm) [inline ccdproc/tests/test_ccdproc.py]: Actually arithmetic no longer works with an ndarray but it has to either be a value or a CCDData object (which can just be an array).   For right now, I'm removing an ndarray as an option and then opening an issue related to that.

File: ccdproc/core.py (hunk 7 above)
Comment (crawfordsm) [inline ccdproc/core.py]: I think we are generally assuming we are starting with a fresh frame, but we can revisit this as needed

**Pull request merged**